Everywhere: Implement explicit struct/class constructor

maddanio opened the pull request:
This makes a user-defined static create function act as an
explicit constructor that is called instead of the implicit one.
The explicit contructor is then the only function which can see
the implicit constructor and uses it to actually create the instance.

awesomekling commented on selfhost/typechecker.jakt:
```diff
@@ -6194,26 +6230,41 @@ struct Typechecker {
+                let explicit_constructor = .get_function(maybe_explicit_constructor_id!)
+                if explicit_constructor.type is ExplicitConstructor {
+                    if .scope_can_access(accessor: scope_id, accessee: explicit_constructor.block.scope_id) {
+                        // fixme: can't we assert this?
```
`maybe_implicit_constructor_id!` has a built-in `VERIFY()`, so can we just drop the `has_value()` check to begin with?

maddanio replied:
I am not sure I understand. The `!` always come after a has value check afaict. there is the question wether we can assert the existence of an explicit constructor, but I am not 100% sure.

maddanio replied:
I.e. in theory an implicit constructor should always exist, but I don't quite understand how external constructors work, so I didnt dare yet :).

AtkinsSJ commented:
This has conflicts. Please fix them. :^)

maddanio commented:
> This has conflicts. Please fix them. :^)

ok, they must be from today, will have a look

maddanio commented:
> This has conflicts. Please fix them. :^)

merged and tests passed

awesomekling commented:
First off, please add tests when introducing a new feature like this. :^)

So, I like the general idea of custom constructors, but the specifics here feel a bit clunky and "magical" with the separation of explicit and implicit constructors.

Here's an alternative proposal:

```
struct Foo {
    field: i64

    ctor {
        // This is the default constructor -> Foo::Foo() in C++
        // It is invoked by doing:
        //     let foo = Foo()
        // You cannot throw in this context.
        field = 0
    }

    ctor(a: i64) {
        // This is an overload constructor -> Foo::Foo(i64 a) in C++
        // It is invoked by doing:
        //     let foo = Foo(a: 123)
        // You cannot throw in this context.
        field = a
    }

    dtor {
        // This is a destructor -> Foo::~Foo() in C++
        // It is invoked when a Foo goes out of scope.
        // You cannot throw in this context.
    }

    function create() -> Foo {
        // This is a factory function that returns a new Foo
        // There's nothing special about it.
        // You can throw in this context.
        return Foo()
    }

    function create(a: i64) -> Foo {
        // This is an overloaded factory function that returns a new Foo
        // There's nothing special about it.
        // You can throw in this context.
        return Foo(a)
    }
}
```


And of course, `struct` constructors are `public` by default, and `class` ones are always `private`.

Basically, modeling constructors and destructors much closer to how they end up in C++. This will make it easier to interface with our existing C++ libraries as well.

maddanio commented:
> First off, please add tests when introducing a new feature like this. :^)

Ah crap, somehow the three test files I had added got lost. Just added them back and pushed. Sorry about that.

> So, I like the general idea of custom constructors, but the specifics here feel a bit clunky and "magical" with the separation of explicit and implicit constructors.
> 
> Here's an alternative proposal:
> 
> ```
> struct Foo {
>     field: i64
> 
>     ctor {
>         // This is the default constructor -> Foo::Foo() in C++
>         // It is invoked by doing:
>         //     let foo = Foo()
>         // You cannot throw in this context.
>         field = 0
>     }
> 
>     ctor(a: i64) {
>         // This is an overload constructor -> Foo::Foo(i64 a) in C++
>         // It is invoked by doing:
>         //     let foo = Foo(a: 123)
>         // You cannot throw in this context.
>         field = a
>     }
> 
>     dtor {
>         // This is a destructor -> Foo::~Foo() in C++
>         // It is invoked when a Foo goes out of scope.
>         // You cannot throw in this context.
>     }
> 
>     function create() -> Foo {
>         // This is a factory function that returns a new Foo
>         // There's nothing special about it.
>         // You can throw in this context.
>         return Foo()
>     }
> 
>     function create(a: i64) -> Foo {
>         // This is an overloaded factory function that returns a new Foo
>         // There's nothing special about it.
>         // You can throw in this context.
>         return Foo(a)
>     }
> }
> ```
> 
> And of course, `struct` constructors are `public` by default, and `class` ones are always `private`.

Ok sure. Then there was a misunderstanding, I had heard from someone else that using `create` and `destruct` had been agreed. So basically you want `ctor` instead and also allow dropping the function keyword and return type and even there brackets (are you sure about the latter one? looks weird to me...). that will require a little parser magic, but not too much.

awesomekling commented:
> > First off, please add tests when introducing a new feature like this. :^)
> 
> Ah crap, somehow the three test files I had added got lost. Just added them back and pushed. Sorry about that.
> 
> > So, I like the general idea of custom constructors, but the specifics here feel a bit clunky and "magical" with the separation of explicit and implicit constructors.
> > Here's an alternative proposal:
> > ```
> > struct Foo {
> >     field: i64
> > 
> >     ctor {
> >         // This is the default constructor -> Foo::Foo() in C++
> >         // It is invoked by doing:
> >         //     let foo = Foo()
> >         // You cannot throw in this context.
> >         field = 0
> >     }
> > 
> >     ctor(a: i64) {
> >         // This is an overload constructor -> Foo::Foo(i64 a) in C++
> >         // It is invoked by doing:
> >         //     let foo = Foo(a: 123)
> >         // You cannot throw in this context.
> >         field = a
> >     }
> > 
> >     dtor {
> >         // This is a destructor -> Foo::~Foo() in C++
> >         // It is invoked when a Foo goes out of scope.
> >         // You cannot throw in this context.
> >     }
> > 
> >     function create() -> Foo {
> >         // This is a factory function that returns a new Foo
> >         // There's nothing special about it.
> >         // You can throw in this context.
> >         return Foo()
> >     }
> > 
> >     function create(a: i64) -> Foo {
> >         // This is an overloaded factory function that returns a new Foo
> >         // There's nothing special about it.
> >         // You can throw in this context.
> >         return Foo(a)
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And of course, `struct` constructors are `public` by default, and `class` ones are always `private`.
> 
> Ok sure. Then there was a misunderstanding, I had heard from someone else that using `create` and `destruct` had been agreed. So basically you want `ctor` instead and also allow dropping the function keyword and return type and even there brackets (are you sure about the latter one? looks weird to me...). that will require a little parser magic, but not too much.

I'm not married to the specific words `ctor` and `dtor`, but we need to call them *something*. We can always change them later. :)

But yes, dropping the `function` keyword makes it clear that these are different from ordinary member functions.

maddanio commented:
> > > First off, please add tests when introducing a new feature like this. :^)
> > 
> > 
> > Ah crap, somehow the three test files I had added got lost. Just added them back and pushed. Sorry about that.
> > > So, I like the general idea of custom constructors, but the specifics here feel a bit clunky and "magical" with the separation of explicit and implicit constructors.
> > > Here's an alternative proposal:
> > > ```
> > > struct Foo {
> > >     field: i64
> > > 
> > >     ctor {
> > >         // This is the default constructor -> Foo::Foo() in C++
> > >         // It is invoked by doing:
> > >         //     let foo = Foo()
> > >         // You cannot throw in this context.
> > >         field = 0
> > >     }
> > > 
> > >     ctor(a: i64) {
> > >         // This is an overload constructor -> Foo::Foo(i64 a) in C++
> > >         // It is invoked by doing:
> > >         //     let foo = Foo(a: 123)
> > >         // You cannot throw in this context.
> > >         field = a
> > >     }
> > > 
> > >     dtor {
> > >         // This is a destructor -> Foo::~Foo() in C++
> > >         // It is invoked when a Foo goes out of scope.
> > >         // You cannot throw in this context.
> > >     }
> > > 
> > >     function create() -> Foo {
> > >         // This is a factory function that returns a new Foo
> > >         // There's nothing special about it.
> > >         // You can throw in this context.
> > >         return Foo()
> > >     }
> > > 
> > >     function create(a: i64) -> Foo {
> > >         // This is an overloaded factory function that returns a new Foo
> > >         // There's nothing special about it.
> > >         // You can throw in this context.
> > >         return Foo(a)
> > >     }
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And of course, `struct` constructors are `public` by default, and `class` ones are always `private`.
> > 
> > 
> > Ok sure. Then there was a misunderstanding, I had heard from someone else that using `create` and `destruct` had been agreed. So basically you want `ctor` instead and also allow dropping the function keyword and return type and even there brackets (are you sure about the latter one? looks weird to me...). that will require a little parser magic, but not too much.
> 
> I'm not married to the specific words `ctor` and `dtor`, but we need to call them _something_. We can always change them later. :)
> 
> But yes, dropping the `function` keyword makes it clear that these are different from ordinary member functions.

ok, implemented now. basically now it is named `ctor` and you have to omit the `function` keyword before it and the return type and are allowed to omit the parentheses entirely if you don't have parameters. last thing I am wondering is should I make the class ctor `throws` implicitly and disallow specifying it and keep it optional for `struct`?

maddanio commented:
forgot to test constructor args and of course failed. fixing

maddanio commented:
> forgot to test constructor args and of course failed. fixing

actually works. I returned the wrong type and return type checking of functions seems fragile, but not going to fix this here

awesomekling commented:
> last thing I am wondering is should I make the class ctor throws implicitly and disallow specifying it and keep it optional for struct?

We cannot throw in our C++ constructors, so let's not support it in Jakt constructors either. At least not initially. We typically use a factory function pattern where something like a `ErrorOr<T> Foo::create()` does all the fallible things, and if successful, returns a newly-constructed `Foo`.

maddanio commented:
Hmm, but if you really want raii then you need to handle the case where the resource acquisition fails. Or would you the always require a factory function for resource like classes like connections or files or the like?

maddanio commented:
So instead of having Socket(adress) you would have Socket::create(address) and male the ctor private (after all what good is an unopened socket)? Seems unnecessarily verbose to me. What i will do is "hide" the implicit throws from the pointer allocation, but i still think constructors should be able to throw.

maddanio commented:
Maybe restrict throw to classes? Feels to me like the right balance?

maddanio commented:
Also disallowing throws on class ctors will make the implementation quite a bit more tricky because in reality it will still throw because in reality it will have to use new. Sure doable just saying :)

maddanio commented:
Unless of course class allocation failure were a panic onstead of a throws :). Anywho, after all I will accept the opinination either way, just trying to make my case :)

maddanio commented:
Ok, after pondering a minute i think i can coerce typechecker and codegen into doing the right thing without too mich complications. Might take a while though because "real life"

maddanio commented:
@awesomekling I worked in your suggestions: no `throws` allowed for `ctor`. the last issue of this PR may be that basically out of implementation lazyness explicit struct ctors will introduce a factory function and also not make the old "implicit" ctor private in the generated c++ code. Calling the factory function and implicit ctor is prevented in jakt via typechecker, but wouldnt be in c++. Would it be important to codegen the explicit struct ctor as a plain c++ constructor to make calling it from c++ easier? I couldn't find any docs regarding that, but it seemed on discord there was some reservations about "deterioating" that interface.

ADKaster commented:
Re: factory functions for structs

My question is, Hlhow important is it to keep the use case of "C++ application or library calling into Jakt library" looking comfortable from the C++ side? 

We'll have to maintain both side of that interface, C++ calling Jakt and Jakt calling C++ (and maybe multiple hops across the boundary in one call stack) for quite some time during any experimentation or transition period for the Serenity codebase. It would be unfortunate if it was difficult or extra verbose to create jakt types in C++ code, but perhaps that's thinking too far ahead?

maddanio commented:
Also for explicit class ctors the factory function will be codegened as `ctor` and the generated implicit constructor still be visible as `create`. Is it a necessity to codegen the explicit ctor as `create` abd make the implicit ctor private in codegen?

maddanio commented:
Well, if there is desire to keep the c++ appearance of jakt types `clean` I can go that extra mile. Might be a while though until I get around to it.

maddanio commented:
Ah, I just figured it will be quite hard to not codegen the explicit struct constructor as a factory function since c++ does not actually allow code in the constructor before the field initializers which the ctors from this pr do (and which i find very convenient). Likely to codegen struct ctors as c++ constructors will require ctors to not have a block but only contain the implicit ctor call

maddanio commented:
But i could still make sure the factory functions are called create and the implicit constructors private in the generated c++ if desired

jntrnr commented:
Thinking through this a bit - there are a few ways jakt can go with constructors.

There's a kind of more traditional C++ approach: you get a default constructor, and you can replace it with a custom one. There is some special treatment for constructions and constructors, for example getting a new keyword just to construct and whatnot.

There's also the Rust approach, which is to not treat constructors and construction any different from any other method. You just make a static method that does the type of construction you want and call it. For example `Foo::new(...)`

I've tried both, and there are advantages/disadvantages to each. Having said that, I think the Rust approach is the simpler of the two. You just make a new static method to build the value how you want. Now that we have default arg values, this can get even cleaner.

As we already support the Rust style today, we could probably stick with that, unless you see advantages it doesn't have.

maddanio commented:
The one thing is you also want to disallow the default constructor in many cases, as it gives you access to private members, which you want to avoud. This is exactly what the ctor does: once it exists you no longer get the default constructor

maddanio commented:
The alternative is to disallow the default constructor whenever there are private fields, but that would break lots of existing code

maddanio commented:
Hmm, so that's it? No explicit ctor and just keave the default ctor with all the private fields spilling out?

maddanio commented:
closing this myself now, as there is no concensus on this, instead we will likely simply hide the implicit ctor if private fields are present
